arch/xtensa/core/xtensa_mmu.c

```diff
@@ -460,13 +806,460 @@ void arch_mem_unmap(void *addr, size_t size)
+
+	/* L2 are done via autofill, so invalidate autofill TLBs
+	 * would refresh the L2 page tables.
+	 *
+	 * L1 will be refreshed during context switch so no need
+	 * to do anything here.
+	 */
+	xtensa_dtlb_autorefill_invalidate_all_sync();
+
 	arch_irq_unlock(key);
 }
+
+#ifdef CONFIG_USERSPACE
+
+static inline uint32_t *alloc_l1_table(void)
+{
+	uint16_t idx;
+
+	for (idx = 0; idx < CONFIG_XTENSA_MMU_NUM_L1_TABLES; idx++) {
+		if (!atomic_test_and_set_bit(l1_page_table_track, idx)) {
+			return (uint32_t *)&l1_page_table[idx];
+		}
+	}
+
+	return NULL;
+}
+
+static uint32_t *dup_table(uint32_t *source_table)
+{
+	uint16_t i, j;
+	uint32_t *dst_table = alloc_l1_table();
+
+	if (!dst_table) {
+		return NULL;
+	}
+
+	for (i = 0; i < XTENSA_L1_PAGE_TABLE_ENTRIES; i++) {
+		uint32_t *l2_table, *src_l2_table;
+
+		if (is_pte_illegal(source_table[i])) {
+			dst_table[i] = Z_XTENSA_MMU_ILLEGAL;
+			continue;
+		}
+
+		src_l2_table = (uint32_t *)(source_table[i] & Z_XTENSA_PTE_PPN_MASK);
+		l2_table = alloc_l2_table();
+		if (l2_table == NULL) {
+			goto err;
+		}
+
+		for (j = 0; j < XTENSA_L2_PAGE_TABLE_ENTRIES; j++) {
+			l2_table[j] =  src_l2_table[j];
+		}
+
+		/* The page table is using kernel ASID because we don't
+		 * user thread manipulate it.
+		 */
+		dst_table[i] = Z_XTENSA_PTE((uint32_t)l2_table, MMU_KERNEL_RING,
+					    Z_XTENSA_MMU_CACHED_WT);
+
+		sys_cache_data_flush_range((void *)l2_table, XTENSA_L2_PAGE_TABLE_SIZE);
+	}
+
+	sys_cache_data_flush_range((void *)dst_table, XTENSA_L1_PAGE_TABLE_SIZE);
+
+	return dst_table;
+
+err:
+	/* TODO: Cleanup failed allocation*/
+	return NULL;
+}
+
+int arch_mem_domain_init(struct k_mem_domain *domain)
+{
+	uint32_t *ptables;
+	k_spinlock_key_t key;
+	int ret;
+
+	/*
+	 * For now, lets just assert if we have reached the maximum number
+	 * of asid we assert.
+	 */
+	__ASSERT(asid_count < (MMU_SHARED_ASID - 1), "Reached maximum of ASID available");
+
+	key = k_spin_lock(&xtensa_mmu_lock);
+	ptables = dup_table(z_xtensa_kernel_ptables);
+
+	if (ptables == NULL) {
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	domain->arch.ptables = ptables;
+	domain->arch.asid = ++asid_count;
+
+	sys_slist_append(&xtensa_domain_list, &domain->arch.node);
+
+	ret = 0;
+
+err:
+	k_spin_unlock(&xtensa_mmu_lock, key);
+
+	return ret;
+}
+
+static int region_map_update(uint32_t *ptables, uintptr_t start,
+			      size_t size, uint32_t ring, uint32_t flags)
+{
+	int ret = 0;
+
+	for (size_t offset = 0; offset < size; offset += CONFIG_MMU_PAGE_SIZE) {
+		uint32_t *l2_table, pte;
+		uint32_t page = start + offset;
+		uint32_t l1_pos = page >> 22;
+		uint32_t l2_pos = Z_XTENSA_L2_POS(page);
+
+		/* Make sure we grab a fresh copy of L1 page table */
+		sys_cache_data_invd_range((void *)&ptables[l1_pos], sizeof(ptables[0]));
+
+		l2_table = (uint32_t *)(ptables[l1_pos] & Z_XTENSA_PTE_PPN_MASK);
+
+		sys_cache_data_invd_range((void *)&l2_table[l2_pos], sizeof(l2_table[0]));
+
+		pte = Z_XTENSA_PTE_RING_SET(l2_table[l2_pos], ring);
```

Comment:
Looking at these lines, they only change the ring and attribute of the PTE, they don't change the PPN. What if the PPN for that l2 PTE was never set (it was previously illegal)? Wouldn't the mapping be incorrect?

Reply:
All virtual-to-physical mappings are the same in all page tables, the address given here is valid, it is not called with arbitrary values.